sdk/python/kfp_tekton/_client.py

```diff
@@ -159,6 +161,61 @@ def __init__(self,
             except FileNotFoundError:
                 logging.info(
                     'Failed to automatically set namespace.', exc_info=False)
+                
+    def wait_for_run_completion(self, run_id: str, timeout: int):
+        """Waits for a run to complete.
+
+        Args:
+          run_id: Run id, returned from run_pipeline.
+          timeout: Timeout in seconds.
+
+        Returns:
+          A run detail object: Most important fields are run and pipeline_runtime.
+
+        Raises:
+          TimeoutError: if the pipeline run failed to finish before the specified timeout.
+        """
+        status = 'Running:'
+        start_time = datetime.datetime.now()
+        if isinstance(timeout, datetime.timedelta):
+            timeout = timeout.total_seconds()
+        is_valid_token = False
+        # Tekton pipelineruns Status
+        # List of Tekton status: https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1/pipelinerun_types.go
+        # List of Tekton error status: https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun.go
+        while (status is None or status.lower()
+               not in ['succeeded',
+                       'failed',
+                       'skipped',
+                       'error',
+                       'completed',
+                       'pipelineruncancelled',
+                       'pipelineruncouldntcancel',
+                       'pipelineruntimeout',
+                       'cancelled',
+                       'stoppedrunfinally',
+                       'cancelledrunfinally',
+                       'invalidtaskresultreference']):
+            try:
+                get_run_response = self._run_api.get_run(run_id=run_id)
+                is_valid_token = True
+            except ApiException as api_ex:
+                # if the token is valid but receiving 401 Unauthorized error
+                # then refresh the token
+                if is_valid_token and api_ex.status == 401:
+                    logging.info('Access token has expired !!! Refreshing ...')
+                    self._refresh_api_client_token()
+                    continue
+                else:
+                    raise api_ex
+            status = get_run_response.run.status
+            elapsed_time = (datetime.datetime.now() -
+                            start_time).total_seconds()
+            logging.info('Waiting for the job to complete...')
+            if elapsed_time > timeout:
+                raise TimeoutError('Run timeout')
+            time.sleep(5)
```

Comment:
nit: even if the pipeline is done, you still need to wait for 5 seconds

Reply:
This is from upstream. Let me fix it here and summit a pr upstream as well.

Reply:
I optimized it better so it returns the status if completed without waiting